e2e/internal/e2e/profile.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2019, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package e2e
+
+import (
+	"testing"
+
+	"github.com/sylabs/singularity/internal/pkg/fakeroot"
+	"github.com/sylabs/singularity/internal/pkg/test/tool/require"
+	"github.com/sylabs/singularity/internal/pkg/util/user"
+)
+
+// SingularityProfile represents a Singularity execution profile.
+type SingularityProfile uint8
+
+const (
+	// UserProfile is the execution profile with a regular user.
+	UserProfile SingularityProfile = iota + 1
+	// RootProfile is the execution profile with root.
+	RootProfile
+	// FakerootProfile is the execution profile with fakeroot.
+	FakerootProfile
+	// UserNamespaceProfile is the execution profile with a user namespace.
+	UserNamespaceProfile
+	// RootUserNamespaceProfile is the execution profile with root and a user namespace.
+	RootUserNamespaceProfile
+)
+
+// Profiles groups all Singularity execution profiles.
+var Profiles = []SingularityProfile{
+	UserProfile,
+	RootProfile,
+	FakerootProfile,
+	UserNamespaceProfile,
+	// NEED FIX: disabled until tests are fixed to run it correctly
+	// RootUserNamespaceProfile,
+}
+
+// CheckRequirements checks and ensures that the corresponding
+// execution profile has all the requirements, the current test
+// is skipped if not.
+func (p SingularityProfile) CheckRequirements(t *testing.T) {
+	switch p {
+	case UserNamespaceProfile, RootUserNamespaceProfile:
+		require.UserNamespace(t)
+	case FakerootProfile:
+		require.UserNamespace(t)
+		// now check that current user has valid mappings
+		// in /etc/subuid and /etc/subgid
+		if _, err := fakeroot.GetIDRange(fakeroot.SubUIDFile, uint32(origUID)); err != nil {
+			t.Skipf("fakeroot configuration error: %s", err)
+		}
+		if _, err := fakeroot.GetIDRange(fakeroot.SubGIDFile, uint32(origUID)); err != nil {
+			t.Skipf("fakeroot configuration error: %s", err)
+		}
+	}
+}
+
+// Match returns if one the provided profiles match the current one.
+func (p SingularityProfile) Match(profiles ...SingularityProfile) bool {
```

Comment:
I'm trying to understand the use case here.

```
if RootProfile.Match(listOfProfiles) {
```

what would the code be trying to do in that case?

I imagine something like this would be easier to read:

```
if listOfProfiles.Match(RootProfile) {
```

but without understanding the problem this is solving I cannot say.

Reply:
If the question that is going to be generally asked (the use case) is "does this list contain this item?" then the second form is clearer:

```
if listOfProfiles.Contains(RootProfile) {
```

If the most common question is going to be "is my current profile in this list?", then this might be clearer:

```
if currentProfile.In(listOfProfiles) {
```

(and notice that I switched from "RootProfile" to "currentProfile" in the second case).

Reply:
The goal of `Match` is to match the current profile against a list of profiles, so yes `In` is fine too. The first form `listOfProfiles.Contains(RootProfile)` would involve much steps to setup and make the things more complex IMO.